pkg/proxy/endpoints.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// EndpointsChangeMap keeps track of endpoint changes for services.
+type EndpointsChangeMap struct {
+	// lock protects items.
+	lock sync.Mutex
+	// hostname is the host where kube-proxy is running.
+	hostname string
+	// items maps a service to is endpointsChange.
+	items map[types.NamespacedName]*endpointsChange
+}
+
+// endpointsChange contains all changes to endpoints that happened since proxy rules were synced.  For a single object,
+// changes are accumulated, i.e. previous is state from before applying the changes,
+// current is state after applying the changes.
+type endpointsChange struct {
+	previous EndpointsMap
+	current  EndpointsMap
+}
+
+// EndpointsMap maps a service to its endpoints.
+type EndpointsMap map[ServicePortName][]Endpoints
+
+// NewEndpointsChangeMap initializes an EndpointsChangeMap
+func NewEndpointsChangeMap(hostname string) EndpointsChangeMap {
+	return EndpointsChangeMap{
+		hostname: hostname,
+		items:    make(map[types.NamespacedName]*endpointsChange),
+	}
+}
+
+// UpdateEndpointMapResult is the updated results after applying endpoints changes.
+type UpdateEndpointMapResult struct {
+	// HCEndpointsLocalIPSize maps an endpoints name to the length of its local IPs.
+	HCEndpointsLocalIPSize map[types.NamespacedName]int
+	// StaleServiceNames identifies if an endpoints service pair is stale.
+	StaleEndpoints map[EndpointServicePair]bool
+	// StaleServiceNames identifies if a service is stale.
+	StaleServiceNames map[ServicePortName]bool
+}
+
+// UpdateEndpointsMap updates endpointsMap base on the given changes.
+func UpdateEndpointsMap(endpointsMap EndpointsMap, changes *EndpointsChangeMap, hostname string) (result UpdateEndpointMapResult) {
+	result.StaleEndpoints = make(map[EndpointServicePair]bool)
+	result.StaleServiceNames = make(map[ServicePortName]bool)
+
+	func() {
```

Comment:
why is this in a func() ?  comment or remove

Reply:
It's a legacy code, we can remove it safely.